### PR TITLE
[Form] Add tel and color types

### PR DESCRIPTION
**Pull request by apetitpa**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #22596
| License       | MIT
| Doc PR        | not provided yet

### CODE REVIEW

File: CONTRIBUTORS.md
Comment (javiereguiluz) [inline CONTRIBUTORS.md]: You can revert this change because that file is autogenerated, so any change will be lost.

Comment (apetitpa) [inline CONTRIBUTORS.md]: I was not aware of this sorry, I'll do it in a minute thank you

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/form.xml
Comment (jvasseur) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/form.xml]: This sounds wrong, a new service shouldn't be already deprecated.

Comment (apetitpa) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/form.xml]: Indeed, I reverted this part thank you

File: src/Symfony/Component/Form/Extension/Core/Type/TelType.php
Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: i definitely prefer `TelephoneType` here :thinking:

Comment (apetitpa) [inline src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: I was quite unsure about what name to use, so I chose this one because it matches the "tel" input name but I have to admit it is less clear

Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: i think fully  qualified is preferred :)

Comment (apetitpa) [inline src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: I'll rename the class then thank you !

Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: @javiereguiluz agree?

Comment (jvasseur) [inline src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: IMO it is better to have a name that match the html input type. Developers already know the html types so keeping the same name help finding the corresponding form type.

Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: imo. telephones matches with tel. It's just a fully qualified reference. Imo. `TelType` is a bad class name :)

**Review by ro0NL — APPROVED** [review]

**Comment by issei-m** [comment]
Looks good😀 
btw would "TelephoneType" be better to be named "TelType" in accordance with type-name of html5 tag? (I'm not familiar with English, so not sure the abbr like "tel" is common used though.)

**Comment by issei-m** [comment]
> would "TelephoneType" be better to be named "TelType" in accordance with type-name of html5 tag?

Oops, I've just missed the review-comment by @ro0NL. Never mind this comment.

**Comment by ro0NL** [comment]
my reasoning; tel === telephone, so TelephoneType is explicit/verbose. In my domain i'd also call this field `$user->telephone`, not `$user->tel`.

File: src/Symfony/Component/Form/Extension/Core/Type/ColorType.php
Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/Core/Type/ColorType.php]: `TextType::class` ?

Comment (apetitpa) [inline src/Symfony/Component/Form/Extension/Core/Type/ColorType.php]: I agree but I have to replace it in other classes too then

Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/Core/Type/ColorType.php]: You should only do this on new classes. We usually avoid doing changes on existing code when it doesn't fix anything or doesn't really improve the situation, in order to avoid useless conflicts when the maintainers are merging lower branches to upper ones and interfering in git history.

File: src/Symfony/Component/Form/Extension/Core/Type/TelType.php
Comment (javiereguiluz) [inline src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: This one is tricky. @apetitpa and @jvasseur are right about using the official HTML type name ... but at the same time, Symfony Forms define a lot of field types that don't match the HTML specification. So maybe it's OK tu use the user-friendly TelephoneType instead of the cryptic TelType? Let's ask a real form expert: @HeahDude what do you think? Thanks!

Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: Put different; Symfony provides `TelephoneType` by implementing `type="tel"`. Point is; we should not name our class to an implementation detail :) IMHO.

Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: For what it's worth: the [misd-service-development/phone-number-bundle](https://github.com/misd-service-development/phone-number-bundle) named it [PhoneNumberType](https://github.com/misd-service-development/phone-number-bundle/blob/master/Form/Type/PhoneNumberType.php) and I actually prefer it over `Tel(ephone)Type`, regardless of what is the html implementation behind.

Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: PhoneNumber, Telephone.. im fine either way :) but yeah.. `PhoneNumberType` is probably even better.. as the 'tele' part is also an implementation detail ;-))

File: src/Symfony/Component/Form/Extension/Core/Type/ColorType.php
Comment (apetitpa) [inline src/Symfony/Component/Form/Extension/Core/Type/ColorType.php]: Thank you for sharing this information, I'll make the change !

File: src/Symfony/Component/Form/Extension/Core/Type/TelType.php
Comment (apetitpa) [inline src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: I like `PhoneNumberType` too but let's wait for @HeahDude comment before I make any change

Comment (Vinorcola) [inline src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: Don't forget to rename your test function also (testTel => testTelephone) ;)

Comment (apetitpa) [inline src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: I will of course, as soon as we have an answer for the class name. Thank you

Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: Hello everyone and thank you @apetitpa for this PR.

Reading the discussion here I understand everyone's point of view :).

>For what it's worth: the misd-service-development/phone-number-bundle (which integrates googlei18n/libphonenumber in Symfony) named it PhoneNumberType and I actually prefer it over Tel(ephone)Type, regardless of what is the html implementation behind.

We actually use this bundle in many applications, and its name sounds consistent to me as it handles phone numbers... with transformers, different widgets, etc.

Adding a simple type here is just a matter of matching the corresponding html input, there is no transformers to ensure the int type or whatever, so I would favor `TelType` as originally proposed by @apetitpa. I think it's consistent with other types that simply extend the `TextType` even if some may think that this is a "bad name" for a class.

I have no strong feeling about this however, if symfony deciders don't like it either I would then propose a simple `PhoneType`. What do you think?

Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: Phonetype sounds good... i like the fact it's semantically more correct when used for fax  and such. I believe using type=tel for that is ok-ish.

Comment (apetitpa) [inline src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: @HeahDude Thank you for your opinion I think `PhoneType` is good replacement if we don't keep `TelType` maybe a final word from @fabpot or @javiereguiluz ?

Comment (dunglas) [inline src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: Usually (for instance in HttpFoundation or the WebLink component), we try to be as close as possible of the underlying standard. I would prefer `TelType` here. Doing this allows to do some nice tricks. For instance, if you want to use [mozilla-services/react-jsonschema-form](https://github.com/mozilla-services/react-jsonschema-form), it will be easier to automatically generate the JSON wanter by this lib from Symfony's metadata if you follow the standard (this libs use official HTML5 types).

Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: Last try.. :)

Given the SF form type refers the HTML spec, it doesnt add much value (semantically) when using forms without rendering.. right? I could simply go with `TextType` as well.

Meaning `UrlType`, `EmailType` also refer HTML spec. However `UrlType` adds value due `FixUrlProtocolListener` but that's is not related to the HTML spec at all.

TLDR; i end up with;

```php
$builder
  ->add('email', TextType::class)
//->add('email', EmailType::class) // better semantics
  ->add('url', UrlType::class) // handles protocols
  ->add('phone', TextType::class)
//->add('phone', TelType::class) // worse semantics
```

I really dont see how a form definition relates to some HTML spec.

Comment (apetitpa) [inline src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: Still no consensus about how the class should be named ? @fabpot @nicolas-grekas @xabbuh

Comment (apetitpa) [reply src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: Hello, I'd like to finish this PR, how should I name the class ? TelType, TelephoneType, PhoneType etc ? @symfony/deciders

Comment (ro0NL) [reply src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: either way no blocker for me :) i understand we just want type="tel" one way or another :+1: (my reasoning still stands though(.

Comment (jvasseur) [reply src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: I think we should go with types that match the html types because that's we we tend to do instead of trying to name them based of semantic of the field (`CheckboxType`, `TextareaType`, `RangeType` are good examples of this)

**Review by ogizanagi — APPROVED** [review]

**Comment by ogizanagi** [comment]
Needs rebase to fix the conflict. (Let's keep `TelephoneType` btw)

**Review by fabpot — APPROVED** [review]

**Comment by apetitpa** [comment]
@ogizanagi @fabpot rebased with 3.4, should I update telephone tests name too ? (testTel => testTelephone)

**Comment by apetitpa** [comment]
@ogizanagi @fabpot rebased with 3.4 (sorry for the delay)

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
@apetitpa Can you have a look at the tests, they seem to be broken by your changes.

**Comment by apetitpa** [comment]
@fabpot It's done, it's green again. Travis 7.1 test will be fixed when merging.

File: src/Symfony/Bundle/FrameworkBundle/Resources/views/Form/color_widget.html.php
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Resources/views/Form/color_widget.html.php]: no other file in this directory has the license header, we should removed

**Review by nicolas-grekas — APPROVED** [review]
with minor comment

File: src/Symfony/Bundle/FrameworkBundle/Resources/views/Form/color_widget.html.php
Comment (apetitpa) [reply src/Symfony/Bundle/FrameworkBundle/Resources/views/Form/color_widget.html.php]: It was suggested by fabbot, but I'll revert the changes ok

**Comment by apetitpa** [comment]
@nicolas-grekas It's done but I still have one concern about the name of the `TelephoneType` class. Should we keep this name or use the original suggested name which was `TelType` as the HTML input type

**Comment by nicolas-grekas** [comment]
If HTML5 is fine with "tel", so am I.

**Comment by apetitpa** [comment]
@nicolas-grekas Done, reverted to `TelType` as agreed

File: src/Symfony/Component/Form/Extension/Core/Type/TelType.php
Comment (nicolas-grekas) [reply src/Symfony/Component/Form/Extension/Core/Type/TelType.php]: @ro0NL I think your reasoning was wrong here when stating this (about "tel"): 

> we should not name our class after an implementation detail

HTML5 is exactly the opposite to an implementation detail: it's a public interface, a standard, the tip of the iceberg. So we should definitely not seek to be more clever than the guys who wrote it (we're not ;) ).

**Comment by nicolas-grekas** [comment]
Thank you @apetitpa.

**Pull request closed without merging**
